### PR TITLE
feat: add listener for Amplify Auth signout to Amplify Auth Adapter

### PR DESCRIPTION
**Pull request by reesscot**
#### Description of changes
This PR listens for Amplify Auth signout events to ensure the Location Store is cleared on signout.

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] `yarn test` passes and tests are updated/added
- [ ] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax
- [ ] _If this change should result in a version bump_, [changeset added](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) (This can be done after creating the PR.) This does not apply to changes made to `docs`, `e2e`, `examples`, or other private packages.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/adapters/createAmplifyAuthAdapter.ts [hunk 1]
```diff
@@ -60,6 +61,13 @@ export const createAmplifyAuthAdapter = (input?: {
...
     listLocations,
     region,
-    registerAuthListener: () => null,
+    registerAuthListener: (onStateChange) => {
+      const cancel = Hub.listen('auth', (data) => {
+        if (data.payload.event === 'signedOut') {
+          onStateChange();
+          cancel();
+        }
+      });
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/adapters/createAmplifyAuthAdapter.ts:70]: Nit: prefer `remove` over `cancel` to align to event listener common naming, e.g. `removeEventListener`

File: packages/react-storage/src/components/StorageBrowser/__tests__/createAmplifyAuthAdapter.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    expect(authAdapter).toHaveProperty('region');
+    expect(authAdapter.region).toEqual('region');
+    expect(authAdapter).toHaveProperty('registerAuthListener');
+  });
+
+  it('should throw if no bucket or region given', () => {
+    jest.spyOn(Amplify, 'getConfig').mockReturnValue({});
+
+    expect(createAmplifyAuthAdapter).toThrow(
+      'Amplify Storage configuration not found. Did you run `Amplify.configure` from your project root?'
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/__tests__/createAmplifyAuthAdapter.spec.tsx:34]: Not a big deal, but maybe we should create a constant from the error message and import here so the test doesn't need to be updated if the error message is updated

File: packages/react-storage/src/components/StorageBrowser/__tests__/createAmplifyAuthAdapter.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    expect(createAmplifyAuthAdapter).toThrow(MISSING_BUCKET_OR_REGION_ERROR);
+  });
+
+  it('should set Hub listener when registerAuthListener is called', () => {
+    const hubListenSpy = jest.spyOn(Hub, 'listen');
+
+    const authAdapter = createAmplifyAuthAdapter();
+    const onStageChange = () => {
+      /* clear location store */
+    };
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/__tests__/createAmplifyAuthAdapter.spec.tsx:44]: Is `onStageChange` called when when you complete a stage or at the start of the next?

File: packages/react-storage/src/components/StorageBrowser/adapters/createAmplifyAuthAdapter.ts [hunk 4]
```diff
@@ -1,10 +1,14 @@
 import { Amplify } from 'aws-amplify';
+import { Hub } from 'aws-amplify/utils';
 import { AuthSession, fetchAuthSession } from 'aws-amplify/auth';
 import { LocationCredentialsProvider } from '@aws-amplify/storage/storage-browser';
 
 import { StorageBrowserAuthAdapter } from './types';
 import { LocationAccess } from '../context/types';
 
+export const MISSING_BUCKET_OR_REGION_ERROR =
+  'Amplify Storage configuration not found. Did you run `Amplify.configure` from your project root?';
```
Comment (calebpollman) [inline packages/react-storage/src/components/StorageBrowser/adapters/createAmplifyAuthAdapter.ts:10]: Appreciate it!

**Review by calebpollman — APPROVED** [review]

File: packages/react-storage/src/components/StorageBrowser/__tests__/createAmplifyAuthAdapter.spec.tsx (hunk 3 above)
Comment (reesscot) [reply packages/react-storage/src/components/StorageBrowser/__tests__/createAmplifyAuthAdapter.spec.tsx]: fixed

**Review by timngyn — APPROVED** [review]

**Pull request merged**